src/main/java/com/minecolonies/coremod/generation/SifterCraftingProvider.java

```diff
@@ -0,0 +1,364 @@
+
+        @NotNull
+        public LootTableBuilder rolls(final int rolls)
+        {
+            this.rolls = rolls;
+            return this;
+        }
+
+        @NotNull
+        public LootTableBuilder bonusRolls(final float bonusRolls)
+        {
+            this.bonusRolls = bonusRolls;
+            return this;
+        }
+
+        @NotNull
+        public LootTableBuilder empty(final int weight)
+        {
+            entries.add(new EntryJson(null, EntryTypeEnum.EMPTY, null, null, false, null, weight, 0));
+            return this;
+        }
+
+        @NotNull
+        public LootTableBuilder item(final Item item, final int weight)
+        {
+            entries.add(new EntryJson(null, EntryTypeEnum.ITEM, item.getRegistryName().toString(), null, false, null, weight, 0));
+            return this;
+        }
+
+        @NotNull
+        public LootTableJson build()
+        {
+            return new LootTableJson(LootTableTypeEnum.GENERIC,
+                    Collections.singletonList(new PoolJson(null, rolls, bonusRolls, entries)));
+        }
+    }
+
+    private class SifterRecipeProvider extends CustomRecipeProvider
+    {
+        public SifterRecipeProvider(@NotNull final DataGenerator generatorIn)
+        {
+            super(generatorIn);
+        }
+
+        @Override
+        protected void registerRecipes(@NotNull final Consumer<IFinishedRecipe> consumer)
+        {
+            for (final Map.Entry<Item, List<SifterMeshDetails>> inputEntry : inputs.entrySet())
+            {
+                for (final SifterMeshDetails mesh : inputEntry.getValue())
+                {
+                    final String name = mesh.getName() + "/" + inputEntry.getKey().getRegistryName().getPath();
+                    final Stream<Item> loot = mesh.getLootTable().getPools().stream()
+                            .flatMap(pool -> pool.getEntries().stream())
+                            .filter(entry -> entry.getType().equals(EntryTypeEnum.ITEM))
+                            .map(entry -> ForgeRegistries.ITEMS.getValue(new ResourceLocation(entry.getName())));
+
+                    CustomRecipeBuilder.create(ModJobs.SIFTER_ID.getPath(), name)
+                            .inputs(Stream.of(new ItemStack(inputEntry.getKey()), new ItemStack(mesh.getMesh())).collect(Collectors.toList()))
+                            .secondaryOutputs(Stream.concat(Stream.of(mesh.getMesh()), loot)
+                                        .map(ItemStack::new)
+                                        .collect(Collectors.toList()))
+                            .lootTable(new ResourceLocation(MOD_ID, "recipes/" + name))
+                            .minBuildingLevel(mesh.getMinBuildingLevel())
+                            .build(consumer);
+                }
+            }
+        }
+    }
+
+    // there is a LootTableProvider but it's a bit more of a pain to work with
+    private class SifterLootTableProvider implements IDataProvider

```

Comment:
I wasn't sure whether it was preferred to use the vanilla LootTableProvider and LootTable.Builder or the LootTableJson from the LDT datagenerators lib; I'm currently using the latter since it allows enumeration, which the other does not, and because the existing block loot generator uses it.  One downside of this is that LootTableJson does not provide a way to name the loot pools.

Reply:
I'm not sure loot pool names is a big deal, and enumeration definitely has it's advantages. I think this is fine.

Reply:
My only hesitation there is that the Forge documentation explicitly states that while names aren't required by vanilla they are required by Forge for the purposes of one of the events that lets mods edit tables.  Having said that, it does appear to actually work as is, so perhaps they're just treating it like a vanilla table and giving it an automatic name.